src/theory/strings/infer_proof_cons.cpp

```diff
@@ -0,0 +1,1043 @@
+            // not necessary, probably first component of equality
+            psb.popStep();
+          }
+          // either equal or rewrites to it
+          std::vector<Node> cexp;
+          if (psb.applyPredTransform(mainEqMain, conc, cexp))
+          {
+            // requires that length success is also true
+            useBuffer = true;
+            Trace("strings-ipc-core") << "...success" << std::endl;
+          }
+          else
+          {
+            Trace("strings-ipc-core") << "...fail" << std::endl;
+          }
+        }
+        else
+        {
+          // should always have given a rule to try above
+          Assert(false) << "No reconstruction rule given for " << infer;
+        }
+      }
+    }
+    break;
+    // ========================== Disequalities
+    case Inference::DEQ_DISL_FIRST_CHAR_STRING_SPLIT:
+    case Inference::DEQ_DISL_STRINGS_SPLIT:
+    {
+      if (conc.getKind() != AND || conc.getNumChildren() != 2
+          || conc[0].getKind() != EQUAL || !conc[0][0].getType().isStringLike()
+          || conc[1].getKind() != EQUAL
+          || conc[1][0].getKind() != STRING_LENGTH)
+      {
+        Trace("strings-ipc-deq") << "malformed application" << std::endl;
+        Assert(false) << "unexpected conclusion " << conc << " for " << infer;
+      }
+      else
+      {
+        Node lenReq =
+            nm->mkNode(GEQ, nm->mkNode(STRING_LENGTH, conc[0][0]), conc[1][1]);
+        Trace("strings-ipc-deq")
+            << "length requirement is " << lenReq << std::endl;
+        if (convertLengthPf(lenReq, ps.d_children, psb))
+        {
+          Trace("strings-ipc-deq") << "...success length" << std::endl;
+          // make the proof
+          std::vector<Node> childrenMain;
+          childrenMain.push_back(lenReq);
+          std::vector<Node> argsMain;
+          argsMain.push_back(nodeIsRev);
+          Node mainConc =
+              psb.tryStep(PfRule::STRING_DECOMPOSE, childrenMain, argsMain);
+          Trace("strings-ipc-deq")
+              << "...main conclusion is " << mainConc << std::endl;
+          useBuffer = (mainConc == conc);
+          Trace("strings-ipc-deq")
+              << "...success is " << useBuffer << std::endl;
+        }
+        else
+        {
+          Trace("strings-ipc-deq") << "...fail length" << std::endl;
+        }
+      }
+    }
+    break;
+    // ========================== Boolean split
+    case Inference::CARD_SP:
+    case Inference::LEN_SPLIT:
+    case Inference::LEN_SPLIT_EMP:
+    case Inference::DEQ_DISL_EMP_SPLIT:
+    case Inference::DEQ_DISL_FIRST_CHAR_EQ_SPLIT:
+    case Inference::DEQ_STRINGS_EQ:
+    case Inference::DEQ_LENS_EQ:
+    case Inference::DEQ_LENGTH_SP:
+    {
+      if (conc.getKind() != OR)
+      {
+        Assert(false) << "Expected OR conclusion for " << infer;
+      }
+      else
+      {
+        ps.d_rule = PfRule::SPLIT;
+        Assert(ps.d_children.empty());
+        ps.d_args.push_back(conc[0]);
+      }
```

Comment:
```suggestion
      Assert(conc.getKind() == OR) << "Expected OR conclusion for " << infer;
      ps.d_rule = PfRule::SPLIT;
      Assert(ps.d_children.empty());
      ps.d_args.push_back(conc[0]);
```

Reply:
This would lead to proof checking failures (due to wrongly applied SPLIT) instead of a debug assertion failure + an application of STRING_TRUST.  This behavior is intentional here.

Reply:
I think that it would be good to define a special function for cases like this that will fail depending on how pedantic proof checking is.

Reply:
This could go in a place like TheoryProofStepBuffer::tryStep. Note that it would be against the philosophy of our infrastructure to allow a bogus ProofNode to be generated (analogous to why we disallow non-well-typed Node).

Reply:
Actually, looking at this again ,this is a fairly specific case of proof reconstruction where we have a target conclusion (conc) and are trying to find its proof.  I think adding generic utilities for this would be overkill at this point, since this is the only place where a kind of "proof inference" is done.